### ISSUE TITLE
Building fails "collect2: error: ld returned 1 exit status" [dev/2.0]

### ISSUE THREAD

**Issue by traverseda** [niftools/nifskope#90]
With 

```

sed -e s/@VERSION@/2.0.dev2/ /home/shared/nifskope/build/README.md.in > /home/shared/nifskope/README.md 
g++ -Wl,-O1 -Wl,-O1,--sort-common,--as-needed,-z,relro -o release/NifSkope GeneratedFiles/.obj/basemodel.o GeneratedFiles/.obj/BlockDXT.o GeneratedFiles/.obj/ColorBlock.o GeneratedFiles/.obj/dds_api.o GeneratedFiles/.obj/DirectDrawSurface.o GeneratedFiles/.obj/Image.o GeneratedFiles/.obj/Stream.o GeneratedFiles/.obj/controllers.o GeneratedFiles/.obj/glcontroller.o GeneratedFiles/.obj/glmarker.o GeneratedFiles/.obj/glmesh.o GeneratedFiles/.obj/glnode.o GeneratedFiles/.obj/glparticles.o GeneratedFiles/.obj/glproperty.o GeneratedFiles/.obj/glscene.o GeneratedFiles/.obj/gltex.o GeneratedFiles/.obj/gltexloaders.o GeneratedFiles/.obj/gltools.o GeneratedFiles/.obj/renderer.o GeneratedFiles/.obj/glview.o GeneratedFiles/.obj/3ds.o GeneratedFiles/.obj/importex.o GeneratedFiles/.obj/obj.o GeneratedFiles/.obj/col.o GeneratedFiles/.obj/kfmmodel.o GeneratedFiles/.obj/kfmxml.o GeneratedFiles/.obj/message.o GeneratedFiles/.obj/nifdelegate.o GeneratedFiles/.obj/nifexpr.o GeneratedFiles/.obj/nifmodel.o GeneratedFiles/.obj/nifproxy.o GeneratedFiles/.obj/nifskope.o GeneratedFiles/.obj/nifskope_ui.o GeneratedFiles/.obj/niftypes.o GeneratedFiles/.obj/nifvalue.o GeneratedFiles/.obj/nifxml.o GeneratedFiles/.obj/nvtristripwrapper.o GeneratedFiles/.obj/qhull.o GeneratedFiles/.obj/settings.o GeneratedFiles/.obj/spellbook.o GeneratedFiles/.obj/animation.o GeneratedFiles/.obj/blocks.o GeneratedFiles/.obj/bounds.o GeneratedFiles/.obj/color.o GeneratedFiles/.obj/flags.o GeneratedFiles/.obj/fo3only.o GeneratedFiles/.obj/havok.o GeneratedFiles/.obj/headerstring.o GeneratedFiles/.obj/light.o GeneratedFiles/.obj/material.o GeneratedFiles/.obj/mesh.o GeneratedFiles/.obj/misc.o GeneratedFiles/.obj/moppcode.o GeneratedFiles/.obj/morphctrl.o GeneratedFiles/.obj/normals.o GeneratedFiles/.obj/optimize.o GeneratedFiles/.obj/sanitize.o GeneratedFiles/.obj/skeleton.o GeneratedFiles/.obj/stringpalette.o GeneratedFiles/.obj/strippify.o GeneratedFiles/.obj/tangentspace.o GeneratedFiles/.obj/texture.o GeneratedFiles/.obj/transform.o GeneratedFiles/.obj/colorwheel.o GeneratedFiles/.obj/fileselect.o GeneratedFiles/.obj/floatedit.o GeneratedFiles/.obj/floatslider.o GeneratedFiles/.obj/groupbox.o GeneratedFiles/.obj/inspect.o GeneratedFiles/.obj/nifcheckboxlist.o GeneratedFiles/.obj/nifeditors.o GeneratedFiles/.obj/nifview.o GeneratedFiles/.obj/refrbrowser.o GeneratedFiles/.obj/uvedit.o GeneratedFiles/.obj/valueedit.o GeneratedFiles/.obj/xmlcheck.o GeneratedFiles/.obj/about_dialog.o GeneratedFiles/.obj/checkablemessagebox.o GeneratedFiles/.obj/settingsdialog.o GeneratedFiles/.obj/version.o GeneratedFiles/.obj/NvTriStrip.o GeneratedFiles/.obj/NvTriStripObjects.o GeneratedFiles/.obj/VertexCache.o GeneratedFiles/.obj/image_DXT.o GeneratedFiles/.obj/image_helper.o GeneratedFiles/.obj/SOIL.o GeneratedFiles/.obj/stb_image_aug.o GeneratedFiles/.obj/qrc_nifskope.o GeneratedFiles/.obj/moc_basemodel.o GeneratedFiles/.obj/moc_glscene.o GeneratedFiles/.obj/moc_gltex.o GeneratedFiles/.obj/moc_icontrollable.o GeneratedFiles/.obj/moc_renderer.o GeneratedFiles/.obj/moc_glview.o GeneratedFiles/.obj/moc_kfmmodel.o GeneratedFiles/.obj/moc_message.o GeneratedFiles/.obj/moc_nifmodel.o GeneratedFiles/.obj/moc_nifproxy.o GeneratedFiles/.obj/moc_nifskope.o GeneratedFiles/.obj/moc_settings.o GeneratedFiles/.obj/moc_spellbook.o GeneratedFiles/.obj/moc_skeleton.o GeneratedFiles/.obj/moc_stringpalette.o GeneratedFiles/.obj/moc_texture.o GeneratedFiles/.obj/moc_colorwheel.o GeneratedFiles/.obj/moc_fileselect.o GeneratedFiles/.obj/moc_floatedit.o GeneratedFiles/.obj/moc_floatslider.o GeneratedFiles/.obj/moc_inspect.o GeneratedFiles/.obj/moc_nifcheckboxlist.o GeneratedFiles/.obj/moc_nifeditors.o GeneratedFiles/.obj/moc_nifview.o GeneratedFiles/.obj/moc_refrbrowser.o GeneratedFiles/.obj/moc_uvedit.o GeneratedFiles/.obj/moc_valueedit.o GeneratedFiles/.obj/moc_xmlcheck.o GeneratedFiles/.obj/moc_about_dialog.o GeneratedFiles/.obj/moc_checkablemessagebox.o GeneratedFiles/.obj/moc_settingsdialog.o   -lGLU -lQt5OpenGL -lQt5Widgets -lQt5Gui -lQt5Xml -lQt5Network -lQt5Core -lGL -lpthread 
GeneratedFiles/.obj/gltex.o: In function `TexCache::find(QString const&, QString const&, QByteArray&)':
gltex.cpp:(.text+0x3a9c): undefined reference to `FSManager::archiveList()'
GeneratedFiles/.obj/nifskope.o: In function `NifSkope::migrateSettings() const::{lambda(QVariant)#3}::operator()(QVariant) const [clone .isra.47] [clone .constprop.94]':
nifskope.cpp:(.text+0x2bda): undefined reference to `FSManager::autodetectArchives()'
GeneratedFiles/.obj/nifskope.o: In function `NifSkope::migrateSettings() const':
nifskope.cpp:(.text+0x5951): undefined reference to `FSManager::get()'
nifskope.cpp:(.text+0x5959): undefined reference to `FSManager::initialize()'
GeneratedFiles/.obj/settings.o: In function `SettingsResources::on_btnArchiveAutoDetect_clicked()':
settings.cpp:(.text+0x2ca9): undefined reference to `FSManager::autodetectArchives()'
GeneratedFiles/.obj/settings.o: In function `SettingsResources::SettingsResources(QWidget*)':
settings.cpp:(.text+0x480a): undefined reference to `FSManager::get()'
GeneratedFiles/.obj/settings.o: In function `SettingsResources::write()':
settings.cpp:(.text+0x721d): undefined reference to `FSArchiveHandler::openArchive(QString const&)'
collect2: error: ld returned 1 exit status
Makefile:681: recipe for target 'release/NifSkope' failed
make: *** [release/NifSkope] Error 1


```

I've tried this branch as well as https://github.com/jonwd7/nifskope/tree/dev/2.0, both fail with the same errors.

qmake-qt5
make clean
make

**Comment by neomonkeus** [comment]
Looks like you are building on *nix can you provide more details on environment
Note : Windows is primary target enviroment, we haven't had a *nix developer so it hasn't been tested.

**Comment by traverseda** [comment]
I can provide any detail you want.

The thousand foot overview is that it's an up to date archlinux system.

```
GNU Make 4.1

QMake version 3.0
Using Qt version 5.5.1 in /usr/lib

gcc (GCC) 5.3.0
```

Let me know if there's anything more specific that would help.

---

As an aside, I'd be pretty happy with a windows build that has support for FO4 meshes. I asked around on IRC if there's a binary anywhere, but results were inconclusive. I'm just not even going to try and set up a build environment under windows.

**Comment by neomonkeus** [comment]
https://github.com/niftools/nifskope/releases

Still will leave this open as nix support is something @niftools aims to support

**Comment by traverseda** [comment]
I'm only seeing a source release, and as previously mentioned I'm not even going to try setting up a windows build environment. Haven't touched windows in years, and I have a hard enough time with the build under *nix.

If there's a windows binary floating around, I would appreciate it. And of course if you want me to gather any more information I will.

**Comment by neomonkeus** [comment]
The link provided has a link in the description to windows binaries (zip). i've updated the description for clarity.

**Comment by traverseda** [comment]
Ahh, thanks.

**Comment by neomonkeus** [comment]
Yeah, sorry for the confusion. we should really migrate the releases from @jonwd7 's repo to the @niftools one.

**Comment by hexabits** [comment]
@traverseda  I said this in IRC, but I would need you to actually try to use Qt Creator for building, because I can't be sure that running qt5-qmake and make on your own isn't the problem. 

The BSA stuff you are getting undefined refs for is in /lib whereas the rest of the program is in /src so I am led to believe that what you are running to build the program isn't enough to set up the environment correctly.

**Comment by DavidMcKenneyBarschall** [comment]
I think this is the error I'm getting when trying to build "2.0dev5" using Qt Creator (4.0.2, Qt 5.7.0) on Ubuntu (14.04 64-bit), except my list of undefined references is even longer.

> GeneratedFiles/.obj/gltex.o: In function `TexCache::find(QString const&, QString const&, QByteArray&)':
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/gl/gltex.cpp:239: undefined reference to`FSManager::archiveList()'
> GeneratedFiles/.obj/nifskope.o: In function `operator()':
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:799: undefined reference to`BSAProxyModel::resetFilter()'
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:1441: undefined reference to `FSManager::autodetectArchives(QString const&)'
> GeneratedFiles/.obj/nifskope.o: In function`NifSkope::openArchive(QString const&)':
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:739: undefined reference to `FSArchiveHandler::openArchive(QString const&)'
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:751: undefined reference to`BSAModel::init()'
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:754: undefined reference to `BSA::fillModel(BSAModel*, QString const&)'
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:773: undefined reference to`BSAProxyModel::setFiletypes(QStringList)'
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:774: undefined reference to `BSAProxyModel::resetFilter()'
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:804: undefined reference to`BSAProxyModel::setFilterByNameOnly(bool)'
> GeneratedFiles/.obj/nifskope.o: In function `NifSkope::migrateSettings() const':
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:1468: undefined reference to`FSManager::get()'
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:1468: undefined reference to `FSManager::initialize()'
> GeneratedFiles/.obj/nifskope.o: In function`NifSkope::NifSkope()':
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:226: undefined reference to `BSAModel::BSAModel(QObject*)'
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:227: undefined reference to`BSAProxyModel::BSAProxyModel(QObject_)'
> GeneratedFiles/.obj/nifskope.o: In function `NifSkope::openArchiveFileString(BSA_, QString const&)':
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:821: undefined reference to `BSA::hasFile(QString const&) const'
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/nifskope.cpp:827: undefined reference to`BSA::fileContents(QString const&, QByteArray&)'
> GeneratedFiles/.obj/settings.o: In function `SettingsResources::on_btnArchiveAutoDetect_clicked()':
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/settings.cpp:760: undefined reference to`FSManager::autodetectArchives(QString const&)'
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/settings.cpp:762: undefined reference to `FSManager::autodetectArchives(QString const&)'
> GeneratedFiles/.obj/settings.o: In function`SettingsResources::SettingsResources(QWidget_)':
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/settings.cpp:498: undefined reference to `FSManager::get()'
> GeneratedFiles/.obj/settings.o: In function`SettingsResources::write()':
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/settings.cpp:571: undefined reference to `FSArchiveHandler::openArchive(QString const&)'
> GeneratedFiles/.obj/material.o: In function`Material::find(QString)':
> build-NifSkope-Desktop_Qt_5_7_0_GCC_64bit-Debug/../nifskope-2.0.dev5/src/material.cpp:116: undefined reference to `FSManager::archiveList()'
> collect2: error: ld returned 1 exit status
> make: *_\* [debug/NifSkope] Error 1
> 12:35:03: The process "/usr/bin/make" exited with code 2.
> Error while building/deploying project NifSkope (kit: Desktop Qt 5.7.0 GCC 64bit)
> When executing step "Make"
> 12:35:03: Elapsed time: 05:20.

**Comment by MiroslavR** [comment]
That's because you're not linking against _fsengine_ lib, which unfortunately does not compile on Linux (apparently due to lib/dds.h requiring windows.h and dxgiformat.h).

From NifSkope.pro:

<pre>
# Dependencies
CONFIG += nvtristrip qhull soil zlib
win32:CONFIG += fsengine
</pre>

**Comment by psi29a** [comment]
I'm running into the same problem. I'll have a look at fsengine in a bit but I'm not hopeful.

@jonwd7 claims on one the issues on his fork that the last time time he tried (with QtCreator) on linux that is compiled just fine. So I'm curious about how he did it as well.

In the mean time, while Nifscope is a C++11 project, I got rid of the min/max macros and replaced the calls with fmin and fmax. That got me to this point. :)

**Comment by psi29a** [comment]
OK, fixed:
https://github.com/niftools/nifskope/pull/101

Thanks @MiroslavR for the hint in the right direction. :)

**Comment by hexabits** [comment]
I believe anything here was resolved with #101 so I am closing.
